lib/common/util.js

```diff
@@ -452,102 +451,228 @@ function shouldRetry(err) {
+ *
+ * @param {object} config - Configuration object.
+ * @param {object=} config.credentials - Credentials object.
+ * @param {string=} config.email - Account email address, required for PEM/P12
+ *     usage.
+ * @param {string=} config.keyFile - Path to a .json, .pem, or .p12 keyfile.
+ * @param {array} config.scopes - Array of scopes required for the API.
+ * @param {function} callback - The callback function.
+ */
+function getAuthClient(config, callback) {
+  var googleAuth = new GoogleAuth();
+
+  if (config.keyFile) {
+    var authClient = new googleAuth.JWT();
+    authClient.keyFile = config.keyFile;
+    authClient.email = config.email;
+    authClient.scopes = config.scopes;
+    addScope(null, authClient);
+  } else if (config.credentials) {
+    googleAuth.fromJSON(config.credentials, addScope);
   } else {
-    authorize = gsa(config);
+    googleAuth.getApplicationDefault(addScope);
   }
 
-  function makeRequest(reqOpts, callback) {
-    var tokenRefreshAttempts = 0;
-    reqOpts.headers = reqOpts.headers || {};
+  function addScope(err, authClient) {
+    if (err) {
+      callback(err);
+      return;
+    }
 
-    if (reqOpts.headers['User-Agent']) {
-      reqOpts.headers['User-Agent'] += '; ' + USER_AGENT;
-    } else {
-      reqOpts.headers['User-Agent'] = USER_AGENT;
+    if (authClient.createScopedRequired && authClient.createScopedRequired()) {
+      authClient = authClient.createScoped(config.scopes);
     }
 
-    function onAuthorizedRequest(err, authorizedReqOpts) {
+    callback(null, authClient);
+  }
+}
+
+util.getAuthClient = getAuthClient;
+
+/**
+ * Authenticate a request by extending its headers object with an access token.
+ *
+ * @param {object} config - Configuration object.
+ * @param {object=} config.credentials - Credentials object.
+ * @param {string=} config.email - Account email address, required for PEM/P12
+ *     usage.
+ * @param {string=} config.keyFile - Path to a .json, .pem, or .p12 keyfile.
+ * @param {array} config.scopes - Array of scopes required for the API.
+ * @param {object} reqOpts - HTTP request options. Its `headers` object is
+ *     created or extended with a valid access token.
+ * @param {function} callback - The callback function.
+ */
+function authorizeRequest(config, reqOpts, callback) {
+  util.getAuthClient(config, function(err, authClient) {
+    if (err) {
+      // google-auth-library returns a "Could not load..." error if it can't get
+      // an access token. However, it's possible an API request doesn't need to
+      // be authenticated, e.g. when downloading a file from a public bucket. We
+      // consider this error a warning, and allow the request to go through
+      // without authorization, relying on the upstream API to return an error
+      // the user would find more helpful, should one occur.
+      if (err.message.indexOf('Could not load') === 0) {
+        callback(null, reqOpts);
+      } else {
+        callback(err);
+      }
+      return;
```

Comment:
In lieu of a return at the end of an `if (err)` block, I just wrap the code after in an `else`. You do this a lot, so it might just be "your style", but I prefer the structural integrity of an else block compared to the flimsy functionality of an undefined `return;` statement. But perhaps that's just me :)